### PR TITLE
Added core plugin structure

### PR DESCRIPTION
**Pull request by diegomedina248**
Added core plugin structure

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Change Log
+All notable changes to BuildSlackNotifier will be documented in this file.
```
Comment (bryant1410) [inline CHANGELOG.md]: This line should be removed

File: README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+* If you **have a feature request** please **open an issue**.
+* If you **found a bug** or **need help** please **check older issues, [FAQ](#faq) and threads on [StackOverflow](http://stackoverflow.com/questions/tagged/BuildSlackNotifier) (Tag 'BuildSlackNotifier') before submitting an issue.**.
+
+Before contribute check the [CONTRIBUTING](CONTRIBUTING.md) file for more info.
+
+If you use **BuildSlackNotifier** in your app We would love to hear about it! Drop us a line on [twitter](https://twitter.com/xmartlabs).
+
+## Author
+
+* [Diego Medina](https://github.com/xmartlabs)
```
Comment (bryant1410) [inline README.md]: Maybe link to your profile, not Xmartlabs'

File: README.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+## Getting involved
+
+* If you **want to contribute** please feel free to **submit pull requests**.
+* If you **have a feature request** please **open an issue**.
+* If you **found a bug** or **need help** please **check older issues, [FAQ](#faq) and threads on [StackOverflow](http://stackoverflow.com/questions/tagged/BuildSlackNotifier) (Tag 'BuildSlackNotifier') before submitting an issue.**.
+
+Before contribute check the [CONTRIBUTING](CONTRIBUTING.md) file for more info.
+
+If you use **BuildSlackNotifier** in your app We would love to hear about it! Drop us a line on [twitter](https://twitter.com/xmartlabs).
```
Comment (bryant1410) [inline README.md]: app We -> app, we

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
+Contributing Guidelines
+--------------------------------------------------
+
+This document provides general guidelines about how to contribute to the project. Keep in mind these important things before you start contributing.
+
+### Asking Questions
```
Comment (bryant1410) [inline CONTRIBUTING.md]: I think we should respect header levels, but who cares

**Review by bryant1410 — CHANGES_REQUESTED** [review]
Left few comments. I avoided to make code style comments as this code won't be maintained a lot.

File: README.md (hunk 2 above)
Comment (diegomedina248) [reply README.md]: yeah that's what I intended :)

**Comment by diegomedina248** [comment]
@bryant1410 if you saw any code style violation, please do comment it here. We should maintain the same standards throughout all of our projects

**Comment by bryant1410** [comment]
> @bryant1410 if you saw any code style violation, please do comment it here. We should maintain the same standards throughout all of our projects

You will regret for sure...

evil

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Short summary of your changes. Explain what have you done and comment the decisions that you may have taken.
+
+### Additional Info:
```
Comment (bryant1410) [inline .github/PULL_REQUEST_TEMPLATE.md]: Here too :smile:

File: CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# Change Log
+
+### [1.0.0](https://github.com/xmartlabs/BuildSlackNotifier/releases/tag/1.0.0)
```
Comment (bryant1410) [inline CHANGELOG.md]: And here

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  @Override
+  public BuildStepMonitor getRequiredMonitorService() {
+    return BuildStepMonitor.NONE;
+  }
+
+  @Extension
+  public static final class DescriptorImpl extends BuildStepDescriptor<Publisher> {
+
```
Comment (bryant1410) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: space

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/SlackURL.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package com.xmartlabs.jenkins.buildslacknotifier;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * Created by diegomedina24 on 11/25/16.
+ */
+public class SlackURL {
+  String url;
```
Comment (bryant1410) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/SlackURL.java]: Couldn't this be private?

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Attachment.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+public class Attachment {
+  private static final String BUILD_NUMBER = "BUILD_NUMBER";
+  private static final String BUILD_URL = "BUILD_URL";
+  private static final String JOB_NAME = "JOB_NAME";
+  private static final String GIT_BRANCH = "GIT_BRANCH";
+  private static final String GIT_COMMIT = "GIT_COMMIT";
+  private static final String FOOTER_TEXT = "Jenkins Build Slack Notifier";
+
+  public AttachmentType color;
```
Comment (bryant1410) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Attachment.java]: Couldn't all these be private?

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/AttachmentField.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package com.xmartlabs.jenkins.buildslacknotifier.request;
+
+import com.google.gson.annotations.SerializedName;
+
+public class AttachmentField {
+  String title;
```
Comment (bryant1410) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/AttachmentField.java]: And these ones

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/AttachmentField.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package com.xmartlabs.jenkins.buildslacknotifier.request;
+
+import com.google.gson.annotations.SerializedName;
+
+public class AttachmentField {
```
Comment (bryant1410) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/AttachmentField.java]: Access can be weaker :smile: please see all Intellij warnings

**Review by bryant1410 — CHANGES_REQUESTED** [review]

**Comment by diegomedina248** [comment]
What a shame @bryant1410, you didn't use the `…` above

**Review by bryant1410 — APPROVED** [review]
Have you tested with these changes? If not, could you do it?

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+import org.kohsuke.stapler.StaplerRequest;
+
+import com.xmartlabs.jenkins.buildslacknotifier.helper.GsonHelper;
+import com.xmartlabs.jenkins.buildslacknotifier.request.Attachment;
+import com.xmartlabs.jenkins.buildslacknotifier.request.Request;
+
+import org.kohsuke.stapler.QueryParameter;
+
+public class BuildSlackNotifier extends Recorder {
+  private List<SlackURL> urls = new ArrayList<>();
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: it should be `final`

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+  @DataBoundConstructor
+  public BuildSlackNotifier(ArrayList<SlackURL> urls) {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: should be `nonNull`

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+  public List<SlackURL> getUrls() {
+    return urls;
+  }
+
+  @Override
+  public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) {
+    try {
+      for (SlackURL url : urls) {
+        sendNotificationToSlack(build, listener, url.getUrl());
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: in my opinion the try/catch should be in inside the `for`

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    public boolean isApplicable(Class<? extends AbstractProject> aClass) {
+      return true;
+    }
+
+    public String getDisplayName() {
+      return "Build Slack Notifier";
+    }
+
+    @Override
+    public boolean configure(StaplerRequest req, JSONObject formData) throws FormException {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: I think that `request` is better than `req`

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/SlackURL.java [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+ */
+public class SlackURL {
+  private String url;
+
+  @DataBoundConstructor
+  public SlackURL(String url) {
+    this.url = url;
+  }
+
+  public void setUrl(String url) {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/SlackURL.java]: is it necessary? shouldn't be `final` ?

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/GsonHelper.java [hunk 17]
```diff
@@ -0,0 +1,20 @@
+package com.xmartlabs.jenkins.buildslacknotifier.helper;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+public class GsonHelper {
+  private static final Gson COMMON_GSON;
+
+  static {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/GsonHelper.java]: remove the static, use inline

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/ResultHelper.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+
+public class ResultHelper {
+  public static String getTitle(Result result, String jobName) {
+    String base = "[" + jobName + "] ";
+    if (result == Result.SUCCESS) {
+      return base + "Successfully built";
+    } else if (result == Result.FAILURE) {
+      return base + "Failed to build";
+    }
+    return "The build completed with some errors, but it wasn't marked as failed";
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/ResultHelper.java:15]: add the result in the result

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Request.java [hunk 19]
```diff
@@ -0,0 +1,12 @@
+package com.xmartlabs.jenkins.buildslacknotifier.request;
+
+import java.util.Collections;
+import java.util.List;
+
+public class Request {
+  private List<Attachment> attachments;
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Request.java]: final

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Request.java [hunk 20]
```diff
@@ -0,0 +1,12 @@
+package com.xmartlabs.jenkins.buildslacknotifier.request;
+
+import java.util.Collections;
+import java.util.List;
+
+public class Request {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Request.java]: final

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Attachment.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.xmartlabs.jenkins.buildslacknotifier.helper.ResultHelper;
+
+import hudson.EnvVars;
+import hudson.model.Result;
+
+public class Attachment {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/request/Attachment.java]: final

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/GsonHelper.java [hunk 22]
```diff
@@ -0,0 +1,20 @@
+package com.xmartlabs.jenkins.buildslacknotifier.helper;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+public class GsonHelper {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/GsonHelper.java]: final

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/SlackURL.java [hunk 23]
```diff
@@ -0,0 +1,23 @@
+package com.xmartlabs.jenkins.buildslacknotifier;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * Created by diegomedina24 on 11/25/16.
+ */
+public class SlackURL {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/SlackURL.java]: final

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.StaplerRequest;
+
+import com.xmartlabs.jenkins.buildslacknotifier.helper.GsonHelper;
+import com.xmartlabs.jenkins.buildslacknotifier.request.Attachment;
+import com.xmartlabs.jenkins.buildslacknotifier.request.Request;
+
+import org.kohsuke.stapler.QueryParameter;
+
+public class BuildSlackNotifier extends Recorder {
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: final

**Review by mirland — CHANGES_REQUESTED** [review]

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/ResultHelper.java (hunk 18 above)
Comment (diegomedina248) [reply src/main/java/com/xmartlabs/jenkins/buildslacknotifier/helper/ResultHelper.java:15]: That's done through the attachment type

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java (hunk 13 above)
Comment (diegomedina248) [reply src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: Can't. For some reason it breaks the plugin (guessing something that Jenkins doesn't like)

**Comment by diegomedina248** [comment]
@bryant1410 I've tested it with these latests changes and seems to be working.

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java [hunk 25]
```diff
@@ -47,12 +47,12 @@ public BuildSlackNotifier(ArrayList<SlackURL> urls) {
 
   @Override
   public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) {
-    try {
-      for (SlackURL url : urls) {
-        sendNotificationToSlack(build, listener, url.getUrl());
+    for (SlackURL url : urls) {
+      try {
+          sendNotificationToSlack(build, listener, url.getUrl());
```
Comment (mirland) [inline src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: indentation

**Review by mirland — APPROVED** [review]

File: src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java (hunk 25 above)
Comment (diegomedina248) [reply src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: shouldn't it be indented with 2 spaces?

Comment (bryant1410) [reply src/main/java/com/xmartlabs/jenkins/buildslacknotifier/BuildSlackNotifier.java]: I'm counting 3 here

**Review by michaelEllisUy — APPROVED** [review]

**Pull request merged**